README.rst
..
    This file is part of Invenio.
    Copyright (C) 2018 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

==================
 Invenio-QA-Tools
==================

.. image:: https://img.shields.io/travis/inveniosoftware/invenio-qa-tools.svg
        :target: https://travis-ci.org/inveniosoftware/invenio-qa-tools

.. image:: https://img.shields.io/coveralls/inveniosoftware/invenio-qa-tools.svg
        :target: https://coveralls.io/r/inveniosoftware/invenio-qa-tools

.. image:: https://img.shields.io/github/tag/inveniosoftware/invenio-qa-tools.svg
        :target: https://github.com/inveniosoftware/invenio-qa-tools/releases

.. image:: https://img.shields.io/pypi/dm/invenio-qa-tools.svg
        :target: https://pypi.python.org/pypi/invenio-qa-tools

.. image:: https://img.shields.io/github/license/inveniosoftware/invenio-qa-tools.svg
        :target: https://github.com/inveniosoftware/invenio-qa-tools/blob/master/LICENSE

Quality assurance tools for Invenio

*This is an experimental developer preview release.*

TODO: Please provide feature overview of module

Further documentation is available on
https://invenio-qa-tools.readthedocs.io/


CHANGES.rst
..
    This file is part of Invenio.
    Copyright (C) 2018 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

Changes
=======

Version 0.1.0 (released TBD)

- Initial public release.


run-tests.sh
#!/usr/bin/env sh
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

pydocstyle invenio_qa_tools tests docs && \
isort -rc -c -df && \
check-manifest --ignore ".travis-*" --ignore ".reana*" && \
sphinx-build -qnNW docs docs/_build/html && \
python setup.py test


tests/conftest.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Pytest configuration."""

from __future__ import absolute_import, print_function

import shutil
import tempfile

import pytest


@pytest.yield_fixture()
def instance_path():
    """Temporary instance path."""
    path = tempfile.mkdtemp()
    yield path
    shutil.rmtree(path)


tests/test_version.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Module tests."""

from __future__ import absolute_import, print_function


def test_version():
    """Test version import."""
    from invenio_qa_tools import __version__
    assert __version__


invenio_qa_tools/api.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Python API of invenio-qa-tools."""

import json

from packaging.utils import canonicalize_name
from packaging.version import Version, parse


def build_package_requirements(package_metadata, **kwargs):
    """Generate package's requirements based on PEP426 metadata.

    Returns a dictionary with following structure from PEP426 package metadata
    definition produced e.g. with by setuptools `dist_info`:

            {package_A:
              {requirement_A: version_specifier_for_requirement_A},
              {requirement_B: version_specifier_for_requirement_B}
            }

    Example output:

            {invenio:
              {sphinx: (>=1.5.1)},
              {isort: (>=4.3.3)}
            }

    If two different version specifiers for same requirement are defined
    both of them are kept (e.g. ["Sphinx (>=1.5.1)", "Sphinx (>1.5)"])

    :param package_metadata:
        String representation of package's JSON metadata as defined by PEP426.
    :type package_metadata: str
    :param **run_requires:
        Should requirements of 'run_requires' key (if present in metadata)
        be taken into account as package's requirements. Defaults to True
    :type **run_requires: bool
    :param **tests_requires:
        Should requirements of 'tests_requires' key (if present in metadata)
        be taken into account as package's requirements. Defaults to True
    :type **tests_requires: bool
    :param **meta_requires:
        Should requirements of 'meta_requires' key (if present in metadata)
        be taken into account as package's requirements. Defaults to True
    :type **meta_requires: bool
    :param **build_requires:
        Should requirements of 'build_requires' key (if present in metadata)
        be taken into account as package's requirements. Defaults to True
    :type **build_requires: bool
    :param **dev_requires:
        Should requirements of 'dev_requires' key (if present in metadata)
        be taken into account as package's requirements. Defaults to True
    :type **dev_requires: bool

    """
    def _get_requirements(key, metadata):
        requirements = metadata.get(key, [])
        if len(requirements) > 1:
            unique_requirements = set()
            for req in requirements:
                if req.get('extra') == 'all':
                    return req.get('requires')
            #     unique_requirements.union(req.get('requires'))
            # return unique_requirements
        elif len(requirements) > 0:  # No 'extras' defined so no need to loop
            return requirements[0].get('requires')
        else:
            return set()

    if not package_metadata:
        raise ValueError('No package metadata provided')

    metadata = json.loads(package_metadata)

    requirements_list = set()  # Later converted to list

    if kwargs.get('run_requires', True):
        run_requires = _get_requirements('run_requires', metadata)
        requirements_list = requirements_list.union(run_requires)

    if kwargs.get('test_requires', True):
        test_requires = _get_requirements('test_requires', metadata)
        requirements_list = requirements_list.union(test_requires)

    if kwargs.get('meta_requires', True):
        meta_requires = _get_requirements('meta_requires', metadata)
        requirements_list = requirements_list.union(meta_requires)

    if kwargs.get('build_requires', True):
        build_requires = _get_requirements('build_requires', metadata)
        requirements_list = requirements_list.union(build_requires)

    if kwargs.get('dev_requires', True):
        dev_requires = _get_requirements('dev_requires', metadata)
        requirements_list = requirements_list.union(dev_requires)

    #  Get package name from metadata
    package_name = metadata['name']

    # Convert requirements_list items to dicts
    # E.g. "Sphinx (>=1.5.1)" to {sphinx: "(>=1.5.1)"}
    reqs = []
    for string in list(requirements_list):
        dep, ver = string.split()  # Split from whitespace characters.
        dep = canonicalize_name(dep)  # Canonicalize name according to PEP426.
        requirement = {dep: ver}  # e.g. {sphinx: "(>=1.5.1)"}
        reqs.append(requirement)

    return {package_name: reqs}


def build_requirements_list(package_list, raise_on_confict=False):
    """Generate list of requirements for a set of packages.

    :param package_list:
    Take following list of dictionaries as input.
    ('version_specifier' is a string defining a version
    specifier for a package's dependency as described by PEP440.)

    [
     {package_A:
          {requirement_A: version_specifier_of_this_pkg_for_requirement_A},
          {requirement_B: version_specifier_of_this_pkg_for_requirement_B}
     },
     {package_B:
          {requirement_A: version_specifier_of_this_pkg_for_requirement_A},
          {requirement_B: version_specifier_of_this_pkg_for_requirement_B},
          {requirement_C: version_specifier_of_this_pkg_for_requirement_C}
     }
    ]

    :param raise_on_confict: Should function continue even if there is a
        conflict in version_specifiers for a requirement. Default is False.
    :type raise_on_conflict: bool

    :return : On successful execution function returns following structure:
    [
     {requirement_A :
          {package_A : [version_specifier(s)_of_this_pkg_for_requirement_A]},
          {package_B : [version_specifier(s)_of_this_pkg_for_requirement_A]}
     },
     {requirement_B:
          {package_A : [version_specifier(s)_of_this_pkg_for_requirement_B]},
          {package_B : [version_specifier(s)_of_this_pkg_for_requirement_B]}
     },
     {requirement_C :
          {package_B : [version_specifier(s)_of_this_pkg_for_requirement_C]}
     }
    ]

    """
    requirements_list = {}

    for package in package_list:
        package_name, deplist = package.popitem()

        for dep in deplist:
            (dep_name, ver), = dep.items()

            if dep_name not in requirements_list:
                templist = []
                templist.append(ver)
                requirements_list[dep_name] = {package_name: templist}
            else:
                if package_name not in requirements_list[dep_name]:
                    requirements_list[dep_name][package_name] = []
                    requirements_list[dep_name][package_name].append(ver)
                else:
                    if not raise_on_confict:
                        requirements_list[dep_name][package_name].append(ver)
                    else:
                        raise ValueError('Package defines multiple '
                                         'version specifiers for '
                                         'same requirement')

            # Add 'version_strings' list to requirement object
            if 'version_specifiers' not in requirements_list[dep_name]:
                requirements_list[dep_name]['version_specifiers'] = []
                requirements_list[dep_name]['version_specifiers'].append(ver)

            if ver not in requirements_list[dep_name]['version_specifiers']:
                requirements_list[dep_name]['version_specifiers'].append(ver)

    return requirements_list


invenio_qa_tools/config.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Quality assurance tools for Invenio."""


invenio_qa_tools/version.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Version information for Invenio-QA-Tools.

This file is imported by ``invenio_qa_tools.__init__``,
and parsed by ``setup.py``.
"""

from __future__ import absolute_import, print_function

__version__ = '0.0.1.dev20180323'


invenio_qa_tools/__init__.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Quality assurance tools for Invenio."""

from __future__ import absolute_import, print_function

import click
import logging
import sys

from invenio_qa_tools import dependencies

from .version import __version__

__all__ = ('__version__',)


DEBUG_LOG_FORMAT = '[%(asctime)s] p%(process)s ' \
                   '{%(pathname)s:%(lineno)d} ' \
                   '%(levelname)s - %(message)s'

LOG_FORMAT = '[%(levelname)s] %(message)s'


class Config(object):
    """Configuration object to share across commands."""

    def __init__(self):
        """Initialize config variables."""


@click.group()
@click.option(
    '--loglevel',
    '-l',
    help='Sets log level',
    type=click.Choice(['DEBUG', 'INFO', 'WARNING']),
    default='WARNING')
@click.pass_context
def cli(ctx, loglevel):
    """Invenio Quality assurance cli tool."""
    logging.basicConfig(
        format=DEBUG_LOG_FORMAT if loglevel == 'DEBUG' else LOG_FORMAT,
        stream=sys.stderr,
        level=loglevel)
    ctx.obj = Config()

cli.add_command(dependencies.check_dependencies)


invenio_qa_tools/dependencies.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2018 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Dependency management related cli commands for invenio-qa-tools."""

from __future__ import absolute_import, print_function

import collections
import json
import os
import sys
from contextlib import contextmanager

import click
import yaml
from setuptools.build_meta import prepare_metadata_for_build_wheel

from invenio_qa_tools.api import build_package_requirements, \
    build_requirements_list


@contextmanager
def suppress_stdout(suppress=True):
    """Suppress output of stdout."""
    if not suppress:
        with open(os.devnull, "w") as devnull:
            old_stdout = sys.stdout
            sys.stdout = devnull
            try:
                yield
            finally:
                sys.stdout = old_stdout
    else:
        yield


@contextmanager
def suppress_stderr(suppress=False):
    """Suppress output of stderr."""
    if suppress:
        with open(os.devnull, "w") as devnull:
            old_stderr = sys.stderr
            sys.stderr = devnull
            try:
                yield
            finally:
                sys.stderr = old_stderr
    else:
        yield


def generate_dist_info(repopath, suppress_output=True, suppress_errors=False):
    """."""
    # 'prepare_metadata_for_build_wheel' requires that root of current working
    #  directory contains 'setup.py' file of the package which build metadata
    # should be generated.
    os.chdir(repopath)
    with suppress_stdout(suppress_output):
        # Stdout is suppressed by default as distutils
        # generate a lot info which is not needed
        with suppress_stderr(suppress_errors):
            dist_info_dir = prepare_metadata_for_build_wheel(repopath)
    os.chdir(os.path.dirname(__file__))
    return dist_info_dir


@click.command(
    'check-dependencies',
    help='TODO')
@click.argument(
    'packages',
    metavar='FILE',
    type=click.Path(exists=True, resolve_path=True),
    nargs=-1)
@click.option('--json', 'output_format', flag_value='json')
@click.option('--yaml', 'output_format', flag_value='yaml')
@click.option('--display-distutils-output', is_flag=True)
@click.option('--suppress-distutils-warnings', is_flag=True)
@click.option('--raise-on-conflict', is_flag=True)
@click.option('--verbose', is_flag=True)
@click.pass_context
def check_dependencies(ctx,
                       packages,
                       output_format,
                       display_distutils_output=False,
                       suppress_distutils_warnings=False,
                       raise_on_conflict=False,
                       verbose=False):
    """."""
    requirements_of_packages = []

    for dir in packages:

        dist_info_dir = generate_dist_info(dir,
                                           display_distutils_output,
                                           suppress_distutils_warnings)

        metadata_path = os.path.join(dir, dist_info_dir, 'metadata.json')
        with click.open_file(metadata_path) as package_metadata:
            metadata = package_metadata.read()
            package_requirements = build_package_requirements(metadata)

        requirements_of_packages.append(package_requirements)

    packages_by_requirement = build_requirements_list(
        requirements_of_packages,
        raise_on_confict=raise_on_conflict)

    if output_format == 'json':
        click.echo(json.dumps(packages_by_requirement))
    elif output_format == 'yaml':
        click.echo(yaml.safe_dump(packages_by_requirement))
    else:

        od = collections.OrderedDict(sorted(
            packages_by_requirement.items(),
            key=lambda i: i[0].lower()))

        for package in od:
            message = click.style(package + '\n')

            conflict = len(
                od[package]['version_specifiers']) > 1

            if conflict:
                message += click.style('\tCONFLICT\n', fg='red')
            else:
                message += click.style('\tOK\n', fg='green')

            for k, v in od[package].items():
                for version in v:
                    if k != 'version_specifiers':
                        message += click.style('\t{} : {}\n'
                                               .format(version, k))

            if not conflict:
                if verbose:
                    click.echo(message)
            else:
                click.echo(message)
